lib/spack/spack/install_test.py

```diff
@@ -110,6 +111,95 @@ def get_test_suite(name):
     return names[0]
 
 
+def package_class(spec):
+    """Return the spec's package class.
+
+    Args:
+        spec (spack.spec.Spec): spec instance
+    """
+    try:
+        cls = spec.package.__class__
+    except AssertionError:
+        try:
+            cls = spec.package_class
+        except spack.repo.UnknownPackageError as e:
+            tty.debug("{0}: Cannot retrieve class from spec: {1}".format(spec.name, str(e)))
+            return None
+
+    return cls
+
+
+def test_functions(spec_or_pkg, add_virtuals=False, names=False):
+    """Grab all non-pass-only test functions (names) associated with the spec or package.
+
+    Args:
+        spec_or_pkg (spack.spec.Spec or spack.package_base.PackageBase): spec or
+            package (class)
+        add_virtuals (bool): True will add any test methods of provided virtuals
+            when a package is provided
+        names (bool): True results in the return of test function names
+
+    Returns: list of test functions or function names
+
+    Raises: ValueError:
+    """
+    if isinstance(spec_or_pkg, spack.spec.Spec):
+        spec = spec_or_pkg
+        cls = package_class(spec)
+        if not cls:
+            tty.debug("Skipping {0}: no package class found".format(spec.name))
+            return []
+    elif isinstance(spec_or_pkg, spack.package_base.PackageBase):
+        pkg = spec_or_pkg
+        cls = pkg.__class__
+        if add_virtuals:
+            methods = []
+            v_names = virtuals(pkg)
+            test_specs = [pkg.spec] + [spack.spec.Spec(v_name) for v_name in sorted(v_names)]
+            for spec in test_specs:
+                methods.extend(test_functions(spec, names=names))
+            return methods
+
+    elif inspect.isclass(spec_or_pkg):
+        cls = spec_or_pkg
+
+    else:
+        raise ValueError("Cannot retrieve test methods for {0}".format(spec_or_pkg))
```

Comment:
Is it worth extracting this part into its own function? If I understand correctly, this gets as input either a spec object or a package object or a package class and returns a package class.

Reply:
Good point.

Reply:
That needed another pass over the logic anyway.

Reply:
Extracted into its own function.